# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/generic/DocAccessible.h
# Commit: 1e75ce1fef67
# Full Hash: 1e75ce1fef67f3139ac4f05993cad693c82c111c
# Author: Eitan Isaacson <eitan@monotonous.org>
# Date: 2020-02-17 09:50:03
# Regressor Bug: 1611060
# File Overlap Count: 1
# Description:
#   Bug 1611060 - Dispatch a11y scroll events for all DOM scrolls. r=Jamie
#   
#   Currently, only documents dispatch scroll events when in reality any
#   element can have scrollable overflows.
#   
# ==============================================================================

diff -r 20b59309d020 -r 1e75ce1fef67 accessible/generic/DocAccessible.h
--- a/accessible/generic/DocAccessible.h	Mon Feb 17 05:37:24 2020 +0200
+++ b/accessible/generic/DocAccessible.h	Mon Feb 17 04:09:34 2020 +0000
@@ -17,7 +17,6 @@
 #include "mozilla/dom/Document.h"
 #include "nsIDocumentObserver.h"
 #include "nsIObserver.h"
-#include "nsIScrollPositionListener.h"
 #include "nsITimer.h"
 
 class nsAccessiblePivot;
@@ -45,7 +44,6 @@
 class DocAccessible : public HyperTextAccessibleWrap,
                       public nsIDocumentObserver,
                       public nsIObserver,
-                      public nsIScrollPositionListener,
                       public nsSupportsWeakReference,
                       public nsIAccessiblePivotObserver {
   NS_DECL_ISUPPORTS_INHERITED
@@ -60,10 +58,6 @@
  public:
   DocAccessible(Document* aDocument, PresShell* aPresShell);
 
-  // nsIScrollPositionListener
-  virtual void ScrollPositionWillChange(nscoord aX, nscoord aY) override {}
-  virtual void ScrollPositionDidChange(nscoord aX, nscoord aY) override;
-
   // nsIDocumentObserver
   NS_DECL_NSIDOCUMENTOBSERVER
 
@@ -384,6 +378,13 @@
    */
   DocAccessibleChild* IPCDoc() const { return mIPCDoc; }
 
+  /**
+   * Notify the document that a DOM node has been scrolled. document will
+   * dispatch throttled accessibility events for scrolling, and a scroll-end
+   * event.
+   */
+  void HandleScroll(nsINode* aTarget);
+
  protected:
   virtual ~DocAccessible();
 
@@ -419,12 +420,6 @@
   void ProcessLoad();
 
   /**
-   * Add/remove scroll listeners, @see nsIScrollPositionListener interface.
-   */
-  void AddScrollListener();
-  void RemoveScrollListener();
-
-  /**
    * Append the given document accessible to this document's child document
    * accessibles.
    */
@@ -577,7 +572,7 @@
    */
   static void ScrollTimerCallback(nsITimer* aTimer, void* aClosure);
 
-  void DispatchScrollingEvent(uint32_t aEventType);
+  void DispatchScrollingEvent(nsINode* aTarget, uint32_t aEventType);
 
   /**
    * Check if an id attribute change affects aria-activedescendant and handle
@@ -607,11 +602,8 @@
    * State and property flags, kept by mDocFlags.
    */
   enum {
-    // Whether scroll listeners were added.
-    eScrollInitialized = 1 << 0,
-
     // Whether the document is a tab document.
-    eTabDocument = 1 << 1
+    eTabDocument = 1 << 0
   };
 
   /**
@@ -623,8 +615,7 @@
 
   Document* mDocumentNode;
   nsCOMPtr<nsITimer> mScrollWatchTimer;
-  uint16_t mScrollPositionChangedTicks;  // Used for tracking scroll events
-  TimeStamp mLastScrollingDispatch;
+  nsDataHashtable<nsPtrHashKey<nsINode>, TimeStamp> mLastScrollingDispatch;
 
   /**
    * Bit mask of document load states (@see LoadState).